src/twisted/trial/_dist/test/test_worker.py

```diff
@@ -86,197 +76,155 @@ class LocalWorkerAMPTests(TestCase):
-
-    def test_runSuccess(self):
+    def test_runSuccess(self) -> None:
         """
         Run a test, and succeed.
         """
-        results = []
-
-        d = self.worker.callRemote(managercommands.AddSuccess, testName=self.testName)
-        d.addCallback(lambda result: results.append(result["success"]))
-        self.pumpTransports()
+        result = self.workerRunTest(pyunitcases.PyUnitTest("test_pass"))
+        assert_that(result, matches_result(successes=equal_to(1)))
 
-        self.assertTrue(results)
-
-    def test_runExpectedFailure(self):
+    def test_runExpectedFailure(self) -> None:
         """
         Run a test, and fail expectedly.
         """
-        results = []
-
-        d = self.worker.callRemote(
-            managercommands.AddExpectedFailure,
-            testName=self.testName,
-            error="error",
-            todo="todoReason",
+        case = skipping.SynchronousStrictTodo("test_todo1")
+        result = self.workerRunTest(case)
+        assert_that(
+            result,
+            matches_result(
+                expectedFailures=equal_to(
+                    [
+                        # Match the strings used in the test we ran.
+                        (case, "expected failure", makeTodo("todo1")),
+                    ]
+                )
+            ),
         )
-        d.addCallback(lambda result: results.append(result["success"]))
-        self.pumpTransports()
 
-        self.assertEqual(self.testCase, self.result.expectedFailures[0][0])
-        self.assertTrue(results)
-
-    def test_runError(self):
+    def test_runError(self) -> None:
         """
         Run a test, and encounter an error.
         """
-        results = []
-        errorClass = fullyQualifiedName(ValueError)
-        d = self.worker.callRemote(
-            managercommands.AddError,
-            testName=self.testName,
-            error="error",
-            errorClass=errorClass,
-            frames=[],
-        )
-        d.addCallback(lambda result: results.append(result["success"]))
-        self.pumpTransports()
-
-        case, failure = self.result.errors[0]
-        self.assertEqual(self.testCase, case)
-        self.assertEqual(failure.type, ValueError)
-        self.assertEqual(failure.value, WorkerException("error"))
-        self.assertTrue(results)
-
-    def test_runErrorWithFrames(self):
-        """
-        L{LocalWorkerAMP._buildFailure} recreates the C{Failure.frames} from
-        the C{frames} argument passed to C{AddError}.
-        """
-        results = []
-        errorClass = fullyQualifiedName(ValueError)
-        d = self.worker.callRemote(
-            managercommands.AddError,
-            testName=self.testName,
-            error="error",
-            errorClass=errorClass,
-            frames=["file.py", "invalid code", "3"],
+        case = pyunitcases.PyUnitTest("test_error")
+        result = self.workerRunTest(case)
+        assert_that(
```

Comment:
at least for me, this embedded structure is not easier to read ... 
but maybe is just my bias for old-school xunit type assertions.

```diff
- self.assertEqual(failure.type, ValueError)
- self.assertEqual(failure.value, WorkerException("error"))
+ self.assertTestRunFailure(self.result, ValueError, WorkerException("error"))
```

For this type of assertion, you only focus on the end/high-level result, and ignore the internal structure of a test case.

It might be an advantage or disadvantage.

I usually write a test in which the internal structure is validated, and then all the other tests only check the high-level values, ignore the internal structure.


With the generic `matches_result` you need to remember the full structure of the test case result, each time you write the test.

Reply:
> at least for me, this embedded structure is not easier to read ...

I don't object to that.  I think a lot of it is about familiarity and experience but it is also true that matcher-style tends to have fewer larger expressions rather than more smaller expressions.

Larger expressions in this style probably often take more practice to become comfortable reading.  The benefit that you get is when multiple parts of the value are wrong, you get to learn about all of them after one test run instead of only the one related to the very first assertion.

testtools allows the best of both worlds with an `expectThat` function which can add a test failure _and_ let the test keep running.  But Twisted decided to go with pyhamcrest instead...

Anyhow, I'm open to improving the readability of these tests before merging.

Reply:
I flattened the matcher in https://github.com/twisted/twisted/pull/11617/commits/6deb9f1bd6a237f3792ecc7e39f5e6d01f457fba - I'm curious if you find that to be any better